src/libraries/System.Net.Security/src/System/Net/Security/SslStream.IO.cs

```diff
@@ -162,13 +162,15 @@ private async Task ReplyOnReAuthenticationAsync<TIOAdapter>(byte[]? buffer, Canc
         private async Task RenegotiateAsync<TIOAdapter>(CancellationToken cancellationToken)
             where TIOAdapter : IReadWriteAdapter
         {
-            if (Interlocked.Exchange(ref _nestedAuth, 1) == 1)
+            if (Interlocked.CompareExchange(ref _nestedAuth, 1, 0) != 0)
             {
+                ObjectDisposedException.ThrowIf(_nestedAuth > 1, this);
```

Comment:
Do we have any tests that try to ReadAsync after calling SslStream.Dispose?

Reply:
I wrote some but I did not include them since I could not reproduce the issue. But I can certainly add them.
Also we have `SslStreamDisposeTest` but none of the tests actually runs because the class is `abstract`. It seems to be this way from the beginning. I plan to look into it as follow-up. Some of the tests needed small adjustments so I did not want to mix it here.

Reply:
I don't mean ones involving any race conditions.  I mean literally:
```C#
sslStream.Dispose();
await sslStream.ReadAsync(...);
```
That should throw an exception today, but presumably this is changing the type of the exception that gets thrown?  We should have a test for that as part of this PR if so.  Or maybe there's another disposal check earlier on in the control flow?

Reply:
I enabled old tests and added new one to have at least some basic coverage